In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium

print('Libraries imported!')

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-1.22.0               |     pyh9f0ad1d_0          63 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          97 KB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.50-py_0
  geopy              conda-forge/noarch::geopy-1.22.0-pyh9f0ad1d_0



geopy-1.22.0         | 63 KB     | ##################################### | 100% 
geographiclib-1.50   | 34 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: 

In [2]:
pip install lxml

     |████████████████████████████████| 5.5MB 25kB/s eta 0:00:0101
Note: you may need to restart the kernel to use updated packages.



### Scraping of table from specified URL

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
tables = pd.read_html(url, header=0,flavor=['lxml'])
tables[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

### Conversion of table to dataframe

In [4]:
df = pd.DataFrame(tables[0])

In [5]:
df.shape

(180, 3)

In [6]:
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

### Removal of rows with no assigned borough

In [7]:
df1 = df[df.Borough != 'Not assigned']

In [8]:
df1.shape

(103, 3)

In [9]:
df1.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [10]:
df1.reset_index(drop=True)

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Vic

### Checking for duplicate postal codes

In [11]:
df2 = df1.groupby(['Postal Code']).groups

In [12]:
len(df2)

103

##### The length of df2 is equal to number of rows of df1, this implies that there are no duplicates.

### Checking for missing neighborhoods

In [13]:
df1.count()

Postal Code     103
Borough         103
Neighborhood    103
dtype: int64

##### The number of values is equal to the number of rows in df1, this implies that here are no missing neighborhoods.

### Size of the resulting dataframe of neighborhoods

In [14]:
df1.shape

(103, 3)

##### The dataframe contains 103 rows.

The Geocoder module did not install, therefore, I decided to get the coordinates from the CSV fil as shown below.

### Reading the csv file with the Postal Code coordinates

In [15]:
df3 = pd.read_csv('Geospatial_Coordinates.csv')
df3.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [16]:
df3.shape

(103, 3)

In [17]:
df4 = pd.merge(df1, df3, how='left', on='Postal Code')

### Resulting dataframe with latitude and longitude

In [18]:
df4

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656

### Exploration and clustering of neighborhoods in Toronto

In [19]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Here is a map of the area under study

In [20]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)
map_toronto

### Creation of new dataframe of Boroughs that contain Toronto

A new data frame with the Boroughs that contain the word Toronto was created in order to work 
with the neighborhoods because Folium does not provide venue information for some of the 
neighborhoods in the complete Toronto area. Therefore, from now on, the work shown is done
with the reduced number of Boroughs.

In [21]:
df5 = df4[df4['Borough'].str.contains('Toronto')].reset_index(drop=True)
df5.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [22]:
df5.shape

(39, 5)

The work will be done with 39 neighborhoods.

### Here is a map of the 39 neighborhoods

In [23]:
# Superimposing neighborhoods in part of Toronto
for lat, lng, borough, neighborhood in zip(df5['Latitude'], df5['Longitude'], df5['Borough'], df5['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Beginning of analysis with Foursquare

In [24]:
CLIENT_ID = '3W1A2N3QRJCNIBMRQEHS5AX0P4JGOKFTZCCBOO0X5CRXCEFF' # your Foursquare ID
CLIENT_SECRET = 'AHW0EUIOGFD1TNT0BGOB3Q1ERYXRZHTC2XKGSIAHTGMZZRFZ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 3W1A2N3QRJCNIBMRQEHS5AX0P4JGOKFTZCCBOO0X5CRXCEFF
CLIENT_SECRET:AHW0EUIOGFD1TNT0BGOB3Q1ERYXRZHTC2XKGSIAHTGMZZRFZ


### Exploring neighborhoods in part of Toronto

In [25]:
def getNearbyVenues(names, latitudes, longitudes):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

The radius was changed to 1000 m since with 500 m Folium dies not make a job finding the clusters. 
For a radius of 500 m Folium assigned 34 neighborhoods to the first cluster and 5 neighborhoods to the other 4 clusters.

In [26]:
LIMIT = 100
radius = 1000
toronto_venues = getNearbyVenues(names=df5['Neighborhood'],
                                   latitudes=df5['Latitude'],
                                   longitudes=df5['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [27]:
print(toronto_venues.shape)
toronto_venues.head()

(3186, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                              Venue  Venue Latitude  Venue Longitude  \
0                  Roselle Desserts       43.653447       -79.362017   
1                     Tandem Coffee       43.653559       -79.361809   
2                    Impact Kitchen       43.656369       -79.356980   
3  The Distillery Historic District       43.650244       -79.359323   
4                   Corktown Common       43.655618       -79.356211   

  Venue Category  
0         Bakery  
1    Coffee Shop  
2     Restaurant  
3  Historic Site  
4           Park

In [28]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                           100   
Brockton, Parkdale Village, Exhibition Place                          100   
Business reply mail Processing Centre, South Ce...                     48   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                    100   
Christie                                                              100   
Church and Wellesley                                                  100   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                            100   
Davisville North                                                      100   
Dufferin, Dovercourt Village                                           65   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                        47   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                         100   
India Bazaar, The Beaches West                                         76   
Kensington Market, Chinatown, Grange Park                             100   
Lawrence Park                                                           9   
Little Portugal, Trinity                                              100   
Moore Park, Summerhill East                                            65   
North Toronto West, Lawrence Park                                      44   
Parkdale, Roncesvalles                                                100   
Queen's Park, Ontario Provincial Government                           100   
Regent Park, Harbourfront                                             100   
Richmond, Adelaide, King                                              100   
Rosedale                                                               21   
Roselawn                                                               25   
Runnymede, Swansea                                                     75   
St. James Town                                                        100   
St. James Town, Cabbagetown                                            37   
Stn A PO Boxes                                                        100   
Studio District                                                       100   
Summerhill West, Rathnelly, South Hill, Forest ...                     77   
The Annex, North Midtown, Yorkville                                   100   
The Beaches                                                            81   
The Danforth West, Riverdale                                          100   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                        100   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                            100   
Brockton, Parkdale Village, Exhibition Place                           100   
Business reply mail Processing Centre, South Ce...                      48   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                     100   
Christie                                                               100   
Church and Wellesley                                                   100   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                             100   
Davisville North                    

### Checking the total number of neighborhoods under study

In [29]:
len(toronto_venues['Neighborhood'].unique())

39

In [30]:
toronto_venues.describe(include='all')

Neighborhood  Neighborhood Latitude  \
count                           3186            3186.000000   
unique                            39                    NaN   
top     The Danforth West, Riverdale                    NaN   
freq                             100                    NaN   
mean                             NaN              43.663404   
std                              NaN               0.019546   
min                              NaN              43.628947   
25%                              NaN              43.648429   
50%                              NaN              43.659526   
75%                              NaN              43.669542   
max                              NaN              43.728020   

        Neighborhood Longitude      Venue  Venue Latitude  Venue Longitude  \
count              3186.000000       3186     3186.000000      3186.000000   
unique                     NaN       1717             NaN              NaN   
top                        NaN  Starbucks             NaN              NaN   
freq                       NaN         56             NaN              NaN   
mean                -79.390193        NaN       43.663198       -79.390271   
std                   0.036804        NaN        0.018947         0.036481   
min                 -79.484450        NaN       43.630680       -79.495322   
25%                 -79.405678        NaN       43.649235       -79.405783   
50%                 -79.384568        NaN       43.658403       -79.386503   
75%                 -79.375418        NaN       43.670304       -79.376668   
max                 -79.293031        NaN       43.733043       -79.282155   

       Venue Category  
count            3186  
unique            274  
top       Coffee Shop  
freq              254  
mean              NaN  
std               NaN  
min               NaN  
25%               NaN  
50%               NaN  
75%               NaN  
max               NaN

In [31]:
toronto_venues.shape

(3186, 7)

### Analysis of each neighborhood

In [32]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# rename column Neighborhood
toronto_onehot.rename(columns={"Neighborhood":"Neighbor"},inplace=True)

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Neighborhood  Accessories Store  Airport  Airport Lounge  \
0  Regent Park, Harbourfront                  0        0               0   
1  Regent Park, Harbourfront                  0        0               0   
2  Regent Park, Harbourfront                  0        0               0   
3  Regent Park, Harbourfront                  0        0               0   
4  Regent Park, Harbourfront                  0        0               0   

   American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
0                    0             0               0             0         0   
1                    0             0               0             0         0   
2                    0             0               0             0         0   
3                    0             0               0             0         0   
4                    0             0               0             0         0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 0   

   Athletics & Sports  Auto Dealership  BBQ Joint  Bagel Shop  Bakery  Bank  \
0                   0                0          0           0       1     0   
1                   0                0          0           0       0     0   
2                   0                0          0           0       0     0   
3                   0                0          0           0       0     0   
4                   0                0          0           0       0     0   

   Bar  Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0    0               0                 0                   0      0   
1    0               0                 0                   0      0   
2    0               0                 0                   0      0   
3    0               0                 0                   0      0   
4    0               0                 0                   0      0   

   Beach Bar  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  Bistro  \
0          0         0           0                   0          0       0   
1          0         0           0                   0          0       0   
2          0         0           0                   0          0       0   
3          0         0           0                   0          0       0   
4          0         0           0                   0          0       0   

   Bookstore  Botanical Garden  Boutique  Brazilian Restaurant  \
0          0                 0         0                     0   
1          0                 0         0                     0   
2          0                 0         0                     0   
3          0                 0         0                     0   
4          0                 0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Buffet  \
0               0        0            0                0       0   
1               0        0            0                0       0   
2               0        0            0                0       0   
3               0        0            0                0       0   
4               0        0            0                0       0   

   Burger Joint  Burrito Place  Bus Stop  Business Service  Butcher  Café  \
0             0              0         0                 0        0     0   
1             0              0         0                 0        0     0   
2             0              0         0                 0        0     0   
3             0              0         0                 0        0     0   
4             0              0         0                 0        0     0   

   Cajun / Creole Restaurant 

In [33]:
toronto_onehot.shape

(3186, 275)

In [34]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped[0:5]

Neighborhood  Accessories Store  \
0                                        Berczy Park               0.00   
1       Brockton, Parkdale Village, Exhibition Place               0.01   
2  Business reply mail Processing Centre, South C...               0.00   
3  CN Tower, King and Spadina, Railway Lands, Har...               0.00   
4                                 Central Bay Street               0.00   

   Airport  Airport Lounge  American Restaurant  Amphitheater  Animal Shelter  \
0   0.0000          0.0000             0.010000           0.0             0.0   
1   0.0000          0.0000             0.010000           0.0             0.0   
2   0.0000          0.0000             0.020833           0.0             0.0   
3   0.0625          0.0625             0.000000           0.0             0.0   
4   0.0000          0.0000             0.010000           0.0             0.0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0         0.02        0.00             0.000000   
1           0.0       0.0         0.01        0.00             0.020000   
2           0.0       0.0         0.00        0.00             0.020833   
3           0.0       0.0         0.00        0.00             0.000000   
4           0.0       0.0         0.03        0.01             0.020000   

   Asian Restaurant  Athletics & Sports  Auto Dealership  BBQ Joint  \
0               0.0                0.00              0.0   0.010000   
1               0.0                0.01              0.0   0.000000   
2               0.0                0.00              0.0   0.020833   
3               0.0                0.00              0.0   0.000000   
4               0.0                0.00              0.0   0.000000   

   Bagel Shop    Bakery  Bank       Bar  Baseball Field  Baseball Stadium  \
0         0.0  0.020000   0.0  0.000000             0.0               0.0   
1         0.0  0.040000   0.0  0.050000             0.0               0.0   
2         0.0  0.020833   0.0  0.020833             0.0               0.0   
3         0.0  0.000000   0.0  0.000000             0.0               0.0   
4         0.0  0.000000   0.0  0.010000             0.0               0.0   

   Basketball Stadium     Beach  Beach Bar  Beer Bar  Beer Store  \
0                0.01  0.010000       0.00      0.03         0.0   
1                0.00  0.000000       0.01      0.01         0.0   
2                0.00  0.020833       0.00      0.00         0.0   
3                0.00  0.000000       0.00      0.00         0.0   
4                0.00  0.000000       0.00      0.01         0.0   

   Belgian Restaurant  Bike Shop    Bistro  Bookstore  Botanical Garden  \
0                 0.0        0.0  0.010000       0.01               0.0   
1                 0.0        0.0  0.000000       0.00               0.0   
2                 0.0        0.0  0.020833       0.00               0.0   
3                 0.0        0.0  0.000000       0.00               0.0   
4                 0.0        0.0  0.000000       0.01               0.0   

   Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0       0.0                   0.0        0.010000   0.0000          0.0   
1       0.0                   0.0        0.010000   0.0100          0.0   
2       0.0                   0.0        0.020833   0.0625          0.0   
3       0.0                   0.0        0.000000   0.0000          0.0   
4       0.0                   0.0        0.010000   0.0000          0.0   

   Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  Bus Stop  \
0             0.01     0.0          0.00       0.000000       0.0   
1             0.00     0.0          0.01       0.000000       0.0   
2             0.00     0.0          0.00       0.020833       0.0   
3             0.00     0.0          0.00       0.000000       0.0   
4             0.02     0.0          0.02       0.020000       0.0   

   Business Service  Butcher   Café

In [35]:
toronto_grouped.shape

(39, 275)

In [36]:
toronto_grouped.describe()

Accessories Store    Airport  Airport Lounge  American Restaurant  \
count          39.000000  39.000000       39.000000            39.000000   
mean            0.000256   0.001603        0.001603             0.009220   
std             0.001601   0.010008        0.010008             0.009911   
min             0.000000   0.000000        0.000000             0.000000   
25%             0.000000   0.000000        0.000000             0.000000   
50%             0.000000   0.000000        0.000000             0.010000   
75%             0.000000   0.000000        0.000000             0.017692   
max             0.010000   0.062500        0.062500             0.040000   

       Amphitheater  Animal Shelter  Antique Shop   Aquarium  Art Gallery  \
count     39.000000       39.000000     39.000000  39.000000    39.000000   
mean       0.000513        0.000256      0.000513   0.000769     0.007399   
std        0.003203        0.001601      0.002235   0.003543     0.009834   
min        0.000000        0.000000      0.000000   0.000000     0.000000   
25%        0.000000        0.000000      0.000000   0.000000     0.000000   
50%        0.000000        0.000000      0.000000   0.000000     0.000000   
75%        0.000000        0.000000      0.000000   0.000000     0.010000   
max        0.020000        0.010000      0.010000   0.020000     0.040000   

       Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
count   39.000000            39.000000         39.000000           39.000000   
mean     0.001026             0.004893          0.003507            0.001734   
std      0.003074             0.007609          0.009731            0.007865   
min      0.000000             0.000000          0.000000            0.000000   
25%      0.000000             0.000000          0.000000            0.000000   
50%      0.000000             0.000000          0.000000            0.000000   
75%      0.000000             0.010000          0.000000            0.000000   
max      0.010000             0.020833          0.040000            0.047619   

       Auto Dealership  BBQ Joint  Bagel Shop     Bakery       Bank  \
count        39.000000  39.000000   39.000000  39.000000  39.000000   
mean          0.000256   0.003887    0.003343   0.021419   0.010066   
std           0.001601   0.008864    0.009446   0.015338   0.018646   
min           0.000000   0.000000    0.000000   0.000000   0.000000   
25%           0.000000   0.000000    0.000000   0.010000   0.000000   
50%           0.000000   0.000000    0.000000   0.020000   0.000000   
75%           0.000000   0.005000    0.000000   0.030000   0.011173   
max           0.010000   0.047619    0.040000   0.070000   0.080000   

             Bar  Baseball Field  Baseball Stadium  Basketball Stadium  \
count  39.000000       39.000000         39.000000           39.000000   
mean    0.016107        0.000594          0.000513            0.001538   
std     0.022069        0.002613          0.003203            0.003655   
min     0.000000        0.000000          0.000000            0.000000   
25%     0.000000        0.000000          0.000000            0.000000   
50%     0.010000        0.000000          0.000000            0.000000   
75%     0.020417        0.000000          0.000000            0.000000   
max     0.070000        0.013158          0.020000            0.010000   

           Beach  Beach Bar   Beer Bar  Beer Store  Belgian Restaurant  \
count  39.000000  39.000000  39.000000   39.000000           39.000000   
mean    0.003009   0.000256   0.006154    0.001224            0.000513   
std     0.009192   0.001601   0.008465    0.003737            0.002235   
min     0.000000   0.000000   0.000000    0.000000            0.000000   
25%     0.000000   0.000000   0.000000    0.000000            0.000000   
50%     0.000000   0.000000   0.000000    0.000000            0.000000   
75%     0.000000   0.000000   0.010000    0.000000            0.000000   
max     

### Five top venues of each neighborhood

In [37]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                 venue  freq
0          Coffee Shop  0.12
1                 Café  0.07
2                Hotel  0.04
3  Japanese Restaurant  0.04
4           Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
         venue  freq
0         Café  0.07
1  Coffee Shop  0.06
2   Restaurant  0.06
3          Bar  0.05
4       Bakery  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0                Park  0.10
1         Coffee Shop  0.06
2         Pizza Place  0.06
3             Brewery  0.06
4  Italian Restaurant  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
             venue  freq
0             Café  0.12
1      Coffee Shop  0.12
2  Harbor / Marina  0.12
3          Dog Run  0.06
4   Airport Lounge  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.08
1       

### Top ten venues for each neighborhood

In [38]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [39]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...                  Park   
3  CN Tower, King and Spadina, Railway Lands, Har...       Harbor / Marina   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café   Japanese Restaurant                 Hotel   
1           Coffee Shop            Restaurant                   Bar   
2               Brewery           Coffee Shop           Pizza Place   
3                  Café           Coffee Shop                  Park   
4                  Café                  Park      Sushi Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Restaurant              Beer Bar                  Park   
1  Furniture / Home Store                Bakery    Tibetan Restaurant   
2    Fast Food Restaurant      Sushi Restaurant    Italian Restaurant   
3                 Airport        Airport Lounge                Garden   
4     Japanese Restaurant           Art Gallery        Cosmetics Shop   

       8th Most Common Venue  9th Most Common Venue  \
0                   Creperie                    Gym   
1                  Gift Shop  Performing Arts Venue   
2                     Bistro      French Restaurant   
3                    Dog Run       Sculpture Garden   
4  Middle Eastern Restaurant                  Hotel   

          10th Most Common Venue  
0                         Bakery  
1  Vegetarian / Vegan Restaurant  
2                     Steakhouse  
3                   Dance Studio  
4                    Yoga Studio

### Cluster neighborhoods

In [40]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 4, 3, 0, 3, 4, 3, 0, 3, 3, 4, 0, 3, 3, 0, 4, 3, 4, 1, 4, 3, 3,
       4, 3, 0, 0, 2, 3, 4, 0, 3, 0, 4, 3, 4, 4, 4, 0, 4], dtype=int32)

This is a much better result of clustering than the one that was made with a radius of 500 m. In
this case, Kmeans was able to separate the neighborhoods in similar clusters.

In [41]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df5

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               3           Coffee Shop   
2  43.657162 -79.378937               3           Coffee Shop   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               4           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0               Theater            Restaurant                  Café   
1      Sushi Restaurant          Burger Joint                  Park   
2   Japanese Restaurant             Gastropub               Theater   
3                  Café            Restaurant             Gastropub   
4                   Pub           Pizza Place                 Beach   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Pub                 Diner                Bakery   
1   Japanese Restaurant      Ramen Restaurant    Italian Restaurant   
2                 Hotel    Italian Restaurant                  Café   
3               Theater    Italian Restaurant    Seafood Restaurant   
4        Breakfast Spot                  Café   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                  Park    Italian Restaurant         Breakfast Spot  
1           Yoga Studio            Restaurant        Bubble Tea Shop  
2                 Diner              Creperie           Burger Joint  
3        Cosmetics Shop   Japanese Restaurant           Concert Hall  
4              Tea Room                Bakery   Caribbean Restaurant

In [42]:
toronto_merged

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                           Regent Park, Harbourfront  43.654260 -79.360636   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
2                            Garden District, Ryerson  43.657162 -79.378937   
3                                      St. James Town  43.651494 -79.375418   
4                                         The Beaches  43.676357 -79.293031   
5                                         Berczy Park  43.644771 -79.373306   
6                                  Central Bay Street  43.657952 -79.387383   
7                                            Christie  43.669542 -79.422564   
8                            Richmond, Adelaide, King  43.650571 -79.384568   
9                        Dufferin, Dovercourt Village  43.669005 -79.442259   
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
11                           Little Portugal, Trinity  43.647927 -79.419750   
12                       The Danforth West, Riverdale  43.679557 -79.352188   
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
15                     India Bazaar, The Beaches West  43.668999 -79.315572   
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
17                                    Studio District  43.659526 -79.340923   
18                                      Lawrence Park  43.728020 -79.388790   
19                                           Roselawn  43.711695 -79.416936   
20                                   Davisville North  43.712751 -79.390197   
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   
22                      High Park, The Junction South  43.661608 -79.464763   
23                  North Toronto West, Lawrence Park  43.715383 -79.405678   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
25                             Parkdale, Roncesvalles  43.648960 -79.456325   
26                                         Davisville  43.704324 -79.388790   
27                     University of Toronto, Harbord  43.662696 -79.400049   
28                                 Runnymede, Swansea  43.651571 -79.484450   
29                        Moore Park, Summerhill East  43.689574 -79.383160   
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
31  Summerhill West, Rathnelly, So

### Here is a map of the different clusters

In [43]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Cluster examination

Cluster 1. Residential zone 1.

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                           Regent Park, Harbourfront           Coffee Shop   
3                                      St. James Town           Coffee Shop   
5                                         Berczy Park           Coffee Shop   
8                            Richmond, Adelaide, King           Coffee Shop   
10  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
13           Toronto Dominion Centre, Design Exchange           Coffee Shop   
16                     Commerce Court, Victoria Hotel           Coffee Shop   
32  CN Tower, King and Spadina, Railway Lands, Har...       Harbor / Marina   
34                                     Stn A PO Boxes           Coffee Shop   
36             First Canadian Place, Underground city           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Theater            Restaurant                  Café   
3                   Café            Restaurant             Gastropub   
5                   Café   Japanese Restaurant                 Hotel   
8                   Café                 Hotel               Theater   
10                  Café                 Hotel   Japanese Restaurant   
13                  Café                 Hotel   Japanese Restaurant   
16                  Café                 Hotel    Seafood Restaurant   
32                  Café           Coffee Shop                  Park   
34                  Café   Japanese Restaurant            Restaurant   
36                  Café                 Hotel               Theater   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                    Pub                 Diner                Bakery   
3                Theater    Italian Restaurant    Seafood Restaurant   
5             Restaurant              Beer Bar                  Park   
8    Japanese Restaurant              Tea Room           Pizza Place   
10                  Park               Theater                   Gym   
13          Concert Hall            Restaurant               Theater   
16   Japanese Restaurant            Restaurant             Gastropub   
32               Airport        Airport Lounge                Garden   
34    Seafood Restaurant             Gastropub                 Hotel   
36   Japanese Restaurant          Concert Hall            Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Park    Italian Restaurant         Breakfast Spot  
3         Cosmetics Shop   Japanese Restaurant           Concert Hall  
5               Creperie                   Gym                 Bakery  
8       Sushi Restaurant          Concert Hall                    Gym  
10               Brewery         Deli / Bodega           Concert Hall  
13             Gastropub          Cocktail Bar     Seafood Restaurant  
16          Concert Hall                Bakery                Theater  
32               Dog Run      Sculpture Garden           Dance Studio  
34                  Park              Beer Bar           Cocktail Bar  
36                  Park         Deli / Bodega     Italian Restaurant

Cluster 2. Study and recreation.

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
18  Lawrence Park                  Café           College Gym   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18             Bookstore                  Park          College Quad   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
18  Gym / Fitness Center           Coffee Shop                 Trail   

   9th Most Common Venue       10th Most Common Venue  
18   Dumpling Restaurant  Eastern European Restaurant

Cluster 3. Recreational area.

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
33  Downtown Toronto               2           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
33     Convenience Store         Grocery Store                  Park   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
33    Athletics & Sports        Breakfast Spot                Bistro   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
33   Filipino Restaurant                  Bank    Japanese Restaurant

Cluster 4. Residential zone 2.

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1   Downtown Toronto               3           Coffee Shop   
2   Downtown Toronto               3           Coffee Shop   
6   Downtown Toronto               3           Coffee Shop   
15      East Toronto               3     Indian Restaurant   
19   Central Toronto               3      Sushi Restaurant   
20   Central Toronto               3           Coffee Shop   
21   Central Toronto               3           Coffee Shop   
23   Central Toronto               3           Coffee Shop   
26   Central Toronto               3    Italian Restaurant   
29   Central Toronto               3    Italian Restaurant   
31   Central Toronto               3           Coffee Shop   
35  Downtown Toronto               3                 Diner   
37  Downtown Toronto               3           Coffee Shop   
38      East Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1       Sushi Restaurant          Burger Joint                  Park   
2    Japanese Restaurant             Gastropub               Theater   
6                   Café                  Park      Sushi Restaurant   
15                  Café           Coffee Shop                 Beach   
19           Coffee Shop              Pharmacy                  Bank   
20    Italian Restaurant            Restaurant                  Café   
21                  Café                  Bank                  Park   
23    Italian Restaurant        Clothing Store                 Diner   
26           Coffee Shop      Sushi Restaurant           Pizza Place   
29         Grocery Store           Coffee Shop                  Park   
31      Sushi Restaurant       Thai Restaurant    Italian Restaurant   
35             Gastropub                  Park                  Café   
37   Japanese Restaurant          Burger Joint                  Park   
38               Brewery           Coffee Shop           Pizza Place   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1    Japanese Restaurant      Ramen Restaurant    Italian Restaurant   
2                  Hotel    Italian Restaurant                  Café   
6    Japanese Restaurant           Art Gallery        Cosmetics Shop   
15  Fast Food Restaurant         Grocery Store                Bakery   
19                   Gym                  Café    Italian Restaurant   
20  Fast Food Restaurant      Sushi Restaurant          Dessert Shop   
21                 Trail          Skating Rink          Burger Joint   
23            Restaurant                  Café    Mexican Restaurant   
26                  Café          Dessert Shop                   Gym   
29        Sandwich Place            Restaurant       Thai Restaurant   
31  Gym / Fitness Center                  Park            Restaurant   
35   Japanese Restaurant            Restaurant         Deli / Bodega   
37            Restaurant      Sushi Restaurant           Yoga Studio   
38  Fast Food Restaurant      Sushi Restaurant    Italian Restaurant   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1                 Yoga Studio            Restaurant        Bubble Tea Shop  
2                       Diner              Creperie           Burger Joint  
6   Middle Eastern Restaurant                 Hotel            Yoga Studio  
15                 Restaurant        Sandwich Place          Burrito Place  
19           Asian Restaurant   Japanese Restaurant                 Bakery  
20                        Gym           Pizza Place              Bookstore  
21           Sushi Restaurant    Italian Restaurant    Japanese Restaurant  
23               Skating Rink   Sporting Goods Shop                   Park  
26  Middle Eastern Restaurant            Restaurant      Indian Restaurant  
29                        Gym            Bagel Shop                   Café  
31              Grocery Store                  Bank                    Gy

Cluster 5. Residential zone 3.

In [48]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4       East Toronto               4           Coffee Shop   
7   Downtown Toronto               4     Korean Restaurant   
9       West Toronto               4                  Café   
11      West Toronto               4                  Café   
12      East Toronto               4      Greek Restaurant   
14      West Toronto               4                  Café   
17      East Toronto               4           Coffee Shop   
22      West Toronto               4                  Café   
24   Central Toronto               4                  Café   
25      West Toronto               4           Coffee Shop   
27  Downtown Toronto               4                  Café   
28      West Toronto               4           Coffee Shop   
30  Downtown Toronto               4                  Café   

            2nd Most Common Venue 3rd Most Common Venue  \
4                             Pub           Pizza Place   
7                            Café           Coffee Shop   
9                     Coffee Shop                  Park   
11                            Bar            Restaurant   
12                           Café           Coffee Shop   
14                    Coffee Shop            Restaurant   
17                            Bar                  Café   
22                    Coffee Shop                   Bar   
24  Vegetarian / Vegan Restaurant           Coffee Shop   
25                           Café        Breakfast Spot   
27                         Bakery           Coffee Shop   
28                           Café                Bakery   
30                            Bar           Coffee Shop   

            4th Most Common Venue   5th Most Common Venue  \
4                           Beach          Breakfast Spot   
7                   Grocery Store            Cocktail Bar   
9                Sushi Restaurant                     Bar   
11  Vegetarian / Vegan Restaurant        Asian Restaurant   
12                            Pub             Pizza Place   
14                            Bar  Furniture / Home Store   
17                        Brewery   Vietnamese Restaurant   
22                Thai Restaurant      Italian Restaurant   
24             Italian Restaurant                  Bakery   
25                    Pizza Place        Sushi Restaurant   
27                            Bar              Restaurant   
28                    Pizza Place                     Pub   
30  Vegetarian / Vegan Restaurant             Art Gallery   

   6th Most Common Venue          7th Most Common Venue 8th Most Common Venue  \
4                   Café            Japanese Restaurant              Tea Room   
7     Mexican Restaurant                    Comedy Club  Ethiopian Restaurant   
9           Gourmet Shop             Italian Restaurant              Pharmacy   
11                Bakery                    Coffee Shop           Pizza Place   
12  Fast Food Restaurant             Italian Restaurant                  Bank   
14                Bakery             Tibetan Restaurant             Gift Shop   
17   American Restaurant                         Bakery                 Diner   
22                  Park                  Deli / Bodega            Restaurant   
24                Museum                    Pizza Place         Grocery Store   
25                   Bar                         Bakery                   Pub   
27             Bookstore  Vegetarian / Vegan Restaurant    Mexican Restaurant   
28    Italian Restaurant                     Restaurant      Sushi Restaurant   
30                Bakery                    Yoga Studio    Mexican Restaurant   

        9th Most Common Venue         10th Most Common Venue  
4                      Bakery           Caribbean Restaurant  
7                 Pizza Place              Indian Restaurant  
9                  Restaurant          Portuguese Restaurant  
11               Cocktail Bar             Italian Restaurant  
12     Furni